src/internal_rules.jl

```diff
@@ -656,19 +656,11 @@ function EnzymeRules.forward(
                     fact.dval[b]
                 end
                 
-                if C <: Array
-                    mul!(dB, dfact, retval, -1, 1)
-                else
-                    tmp = dfact.U * retval
-
-                    dB .-= dfact.L * tmp
-
-                    # if mul! was implemented for LU, this would be faster
-                    # mul!(dB, dfact.L, tmp, -1, 1)
-                end
+                tmp = dfact.U * retval
```

Comment:
I'd honestly prefer to leave the support for Array here, if anything because it's what I'd like to exist as an api for cholesky -- but also this is much more efficient for arrays and I'd be interested in adding such a method in the future.

Reply:
But this branch can never be hit, so why should it be kept? There's no method `ldiv!(::Matrix, ...)` (more general `Array`s can definitely not work).

Reply:
At minimum can you leave the comment 

```
# if mul! was implemented for LU, this would be faster
# mul!(dB, dfact.L, tmp, -1, 1)
```

This is the general fast and correct rule, which doesn't exist as a method, unfortunately doesn't exist as a method for diagonal matricies atm (which I hope gets added).

Reply:
I removed the comment because actually this line is already used (just one line below). `mul!` is already defined for triangular and diagonal matrices:
```julia
julia> methods(mul!, Tuple{AbstractVecOrMat, Diagonal, AbstractVecOrMat, Number, Number})
# 6 methods for generic function "mul!" from LinearAlgebra:
 [1] mul!(C::AbstractVector, A::Union{Bidiagonal, Diagonal, SymTridiagonal, Tridiagonal}, B::AbstractVector, alpha::Number, beta::Number)
     @ ~/.asdf/installs/julia/1.10.0/share/julia/stdlib/v1.10/LinearAlgebra/src/bidiag.jl:425
 [2] mul!(C::AbstractMatrix, A::Union{Bidiagonal, Diagonal, SymTridiagonal, Tridiagonal}, B::LinearAlgebra.AbstractTriangular, alpha::Number, beta::Number)
     @ ~/.asdf/installs/julia/1.10.0/share/julia/stdlib/v1.10/LinearAlgebra/src/special.jl:115
 [3] mul!(C::AbstractMatrix, A::Union{Bidiagonal, Diagonal, SymTridiagonal, Tridiagonal}, B::Union{Bidiagonal, Diagonal, SymTridiagonal, Tridiagonal}, alpha::Number, beta::Number)
     @ ~/.asdf/installs/julia/1.10.0/share/julia/stdlib/v1.10/LinearAlgebra/src/bidiag.jl:429
 [4] mul!(C::AbstractMatrix, A::Union{Bidiagonal, Diagonal, SymTridiagonal, Tridiagonal}, B::AbstractMatrix, alpha::Number, beta::Number)
     @ ~/.asdf/installs/julia/1.10.0/share/julia/stdlib/v1.10/LinearAlgebra/src/bidiag.jl:427
 [5] mul!(C::AbstractMatrix, A::Union{Bidiagonal, Diagonal, SymTridiagonal, Tridiagonal}, B::AbstractVector, alpha::Number, beta::Number)
     @ ~/.asdf/installs/julia/1.10.0/share/julia/stdlib/v1.10/LinearAlgebra/src/bidiag.jl:426
 [6] mul!(C::AbstractMatrix, A::AbstractVecOrMat, B::AbstractVecOrMat, α::Number, β::Number)
     @ ~/.asdf/installs/julia/1.10.0/share/julia/stdlib/v1.10/LinearAlgebra/src/matmul.jl:263

julia> methods(mul!, Tuple{AbstractVecOrMat, LinearAlgebra.AbstractTriangular, AbstractVecOrMat, Number, Number})
# 6 methods for generic function "mul!" from LinearAlgebra:
 [1] mul!(C::AbstractVector, A::LinearAlgebra.AbstractTriangular, B::AbstractVector, alpha::Number, beta::Number)
     @ ~/.asdf/installs/julia/1.10.0/share/julia/stdlib/v1.10/LinearAlgebra/src/triangular.jl:724
 [2] mul!(C::AbstractMatrix, A::LinearAlgebra.AbstractTriangular, B::Union{Bidiagonal, Diagonal, SymTridiagonal, Tridiagonal}, alpha::Number, beta::Number)
     @ ~/.asdf/installs/julia/1.10.0/share/julia/stdlib/v1.10/LinearAlgebra/src/special.jl:113
 [3] mul!(C::AbstractMatrix, A::LinearAlgebra.AbstractTriangular, B::LinearAlgebra.AbstractTriangular, alpha::Number, beta::Number)
     @ ~/.asdf/installs/julia/1.10.0/share/julia/stdlib/v1.10/LinearAlgebra/src/triangular.jl:736
 [4] mul!(C::AbstractMatrix, A::LinearAlgebra.AbstractTriangular, B::AbstractMatrix, alpha::Number, beta::Number)
     @ ~/.asdf/installs/julia/1.10.0/share/julia/stdlib/v1.10/LinearAlgebra/src/triangular.jl:736
 [5] mul!(C::AbstractMatrix, A::LinearAlgebra.AbstractTriangular, B::AbstractVector, alpha::Number, beta::Number)
     @ ~/.asdf/installs/julia/1.10.0/share/julia/stdlib/v1.10/LinearAlgebra/src/triangular.jl:724
 [6] mul!(C::AbstractMatrix, A::AbstractVecOrMat, B::AbstractVecOrMat, α::Number, β::Number)
     @ ~/.asdf/installs/julia/1.10.0/share/julia/stdlib/v1.10/LinearAlgebra/src/matmul.jl:263
```

Reply:
oh that's cool. If it builds/runs that's great. I somehow remember it failing, so didn't use